core-rust/state-manager/src/protocol/protocol_update.rs

```diff
@@ -0,0 +1,489 @@
+};
+
+pub trait ProtocolUpdaterFactory {
+    fn supports_protocol_version(&self, protocol_version_name: &str) -> bool;
+
+    fn updater_for(
+        &self,
+        protocol_version_name: &str,
+        store: Arc<StateLock<StateManagerDatabase>>,
+    ) -> Box<dyn ProtocolUpdater>;
+}
+
+/// Protocol update consists of two events:
+/// 1) Updating the current (state computer) configuration ("transaction processing rules").
+///    This includes: transaction validation, execution configuration, etc
+/// 2) Executing arbitrary state updates against the current database state.
+///    While the abstraction is quite flexible, the only concrete implementation at the moment
+///    only modifies the state through committing system transactions (e.g. substate flash).
+pub trait ProtocolUpdater {
+    fn protocol_version_name(&self) -> String;
+    fn state_computer_configurator(&self) -> StateComputerConfigurator;
+    fn state_update_executor(&self) -> Box<dyn StateUpdateExecutor>;
+}
+
+#[derive(Clone, Debug)]
+pub struct StateComputerConfigurator {
+    pub network: NetworkDefinition,
+    pub logging_config: LoggingConfig,
+    pub validation_config: ValidationConfig,
+    pub costing_parameters: CostingParameters,
+}
+
+impl StateComputerConfigurator {
+    pub fn default(network: NetworkDefinition) -> StateComputerConfigurator {
+        let network_id = network.id;
+        StateComputerConfigurator {
+            network,
+            logging_config: LoggingConfig::default(),
+            validation_config: ValidationConfig::default(network_id),
+            costing_parameters: CostingParameters::default(),
+        }
+    }
+}
+
+impl StateComputerConfigurator {
+    pub fn ledger_transaction_validator(&self) -> LedgerTransactionValidator {
+        LedgerTransactionValidator::default_from_validation_config(self.validation_config)
+    }
+
+    pub fn user_transaction_validator(&self) -> NotarizedTransactionValidator {
+        NotarizedTransactionValidator::new(self.validation_config)
+    }
+
+    pub fn validation_config(&self) -> ValidationConfig {
+        self.validation_config
+    }
+
+    pub fn execution_configurator(&self, no_fees: bool) -> ExecutionConfigurator {
+        let mut costing_parameters = self.costing_parameters;
+        if no_fees {
+            costing_parameters.execution_cost_unit_price = Decimal::ZERO;
+            costing_parameters.finalization_cost_unit_price = Decimal::ZERO;
+            costing_parameters.state_storage_price = Decimal::ZERO;
+            costing_parameters.archive_storage_price = Decimal::ZERO;
+        }
+        ExecutionConfigurator::new(&self.network, &self.logging_config, costing_parameters)
+    }
+}
+
+pub trait StateUpdateExecutor {
+    /// Executes these state updates associated with the given protocol version
+    /// that haven't yet been applied
+    /// (hence "remaining", e.g. if node is restarted mid-protocol update).
+    fn execute_remaining_state_updates(&self);
```

Comment:
Feels a bit of code smell that the "remaining logic" part is hidden behind a trait rather than implemeted as explicit logic.